ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/storage/migration/PrivateStorageMigrationTransactionProcessor.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.ethereum.privacy.storage.migration;
+
+import org.hyperledger.besu.ethereum.chain.Blockchain;
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+import org.hyperledger.besu.ethereum.core.Hash;
+import org.hyperledger.besu.ethereum.core.MutableWorldState;
+import org.hyperledger.besu.ethereum.debug.TraceOptions;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSchedule;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSpec;
+import org.hyperledger.besu.ethereum.privacy.PrivateStateRootResolver;
+import org.hyperledger.besu.ethereum.privacy.PrivateTransaction;
+import org.hyperledger.besu.ethereum.privacy.PrivateTransactionProcessor;
+import org.hyperledger.besu.ethereum.privacy.PrivateTransactionSimulatorResult;
+import org.hyperledger.besu.ethereum.vm.BlockHashLookup;
+import org.hyperledger.besu.ethereum.vm.DebugOperationTracer;
+import org.hyperledger.besu.ethereum.worldstate.WorldStateArchive;
+
+import java.util.Optional;
+
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+
+public class PrivateStorageMigrationTransactionProcessor {
+
+  private final Blockchain blockchain;
+  private final ProtocolSchedule<?> protocolSchedule;
+  private final WorldStateArchive publicWorldStateArchive;
+  private final WorldStateArchive privateWorldStateArchive;
+  private final PrivateStateRootResolver privateStateRootResolver;
+
+  public PrivateStorageMigrationTransactionProcessor(
+      final Blockchain blockchain,
+      final ProtocolSchedule<?> protocolSchedule,
+      final WorldStateArchive publicWorldStateArchive,
+      final WorldStateArchive privateWorldStateArchive,
+      final PrivateStateRootResolver privateStateRootResolver) {
+    this.blockchain = blockchain;
+    this.protocolSchedule = protocolSchedule;
+    this.publicWorldStateArchive = publicWorldStateArchive;
+    this.privateWorldStateArchive = privateWorldStateArchive;
+    this.privateStateRootResolver = privateStateRootResolver;
+  }
+
+  public Optional<PrivateTransactionSimulatorResult> process(
+      final String privacyGroupId,
+      final PrivateTransaction privateTransaction,
+      final BlockHeader header) {
+    if (header == null) {
+      return Optional.empty();
+    }
+
+    final MutableWorldState publicWorldState =
```

Comment:
When we are simulating this private transaction we need to use the public state at the end of the public Tx just before the PMT.